From f8f7c5f984d4148be99ad06f42dd6c8350d25c04 Mon Sep 17 00:00:00 2001
From: Pier Angelo Vendrame <pierov@torproject.org>
Date: Tue, 3 Sep 2024 19:55:05 +0200
Subject: [PATCH] BB 42773: Replace ~ with the original home.

In Bug 93141, Mozilla started sending users to their home when they type
~ in the URL bar.
On Linux, we change $HOME for various reason, therefore you would be
redirected to the spoofed home directory when typing ~.
So, we check if the original home directory is known, and use that,
instead.
---
 docshell/base/URIFixup.sys.mjs | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/docshell/base/URIFixup.sys.mjs b/docshell/base/URIFixup.sys.mjs
index 0fddbbba40..f6ff988801 100644
--- a/docshell/base/URIFixup.sys.mjs
+++ b/docshell/base/URIFixup.sys.mjs
@@ -924,6 +924,10 @@ function fileURIFixup(uriString) {
   } else {
     // UNIX: Check if it starts with "/" or "~".
     attemptFixup = /^[~/]/.test(uriString);
+    const originalHome = Services.env.get("BB_ORIGINAL_HOME");
+    if (originalHome && (uriString === "~" || uriString.startsWith("~/"))) {
+      path = originalHome + path.substring(1);
+    }
   }
   if (attemptFixup) {
     try {
-- 
2.50.1

